### DISCUSSION TITLE
Spec and Module Naming Conventions

### DISCUSSION THREAD

**Discussion by jerielng** [reactwg/react-native-new-architecture#124] [Documentation]
Hey all, I might be getting my wires crossed, but looking through this [page](https://reactnative.dev/docs/the-new-architecture/pillars-turbomodules), I was wondering if we could get some clarifications on some of the naming conventions involved in migrating to a Turbo Module.

In the examples, I'm seeing different instances of how the module is named, from the Spec file to the actual module name. Do we have any guidance for unifying the naming and differentiating between the nuances?

Is my understanding correct:
1. The actual JS/TS spec file itself should follow the `Native<module>.ts` convention strictly. This is for Codegen to work, correct?
2. Further, the name of the module in the example is `RTNCalculator`, but I see mentions in the Codegen portion that it's referred to as `RTNCalculatorSpec`. In some places, I'm seeing it referred to as `NativeCalculatorSpec` to match the `.ts` file.

For the purposes of pattern matching with this example, what's the difference between `NativeCalculatorSpec` and `RTNCalculatorSpec` here? Seems like Codegen generates the artifacts as `RTNCalculatorSpec` but it's later referenced in the native code as `NativeCalculatorSpec`?

I wanted to double check if I was missing a link here?

**Comment by cortinico** [comment]
Hi @jerielng 

> The actual JS/TS spec file itself should follow the Native<module>.ts convention strictly. This is for Codegen to work, correct?

That's correct. It's a requirement of the codegen.

> Further, the name of the module in the example is RTNCalculator, but I see mentions in the Codegen portion that it's referred to as RTNCalculatorSpec. In some places, I'm seeing it referred to as NativeCalculatorSpec to match the .ts file.
> For the purposes of pattern matching with this example, what's the difference between NativeCalculatorSpec and RTNCalculatorSpec here? Seems like Codegen generates the artifacts as RTNCalculatorSpec but it's later referenced in the native code as NativeCalculatorSpec?

Yup I agree that this is confusing:
- `RTNCalculatorSpec` is the name we used for the codegen library. The codegen needs to be invoked with  a library name fo some sort. It can be anything (like `mymodules`) and will be used to codegen library wide information (e.g. the folder for the header will be `mymodules/...`). We used `RTNCalculatorSpec` for convience but that's probably confusing in the examples.
- `NativeXYZSpec` is the class that codegen generates from a spec file found as `NativeXYZ.js`. We generate both Android and iOS classes that you have to extent/implement, that's why you find them referenced in the examples.

**Reply by jerielng** [reply]
Thanks for providing the distinction! That was really helpful!
